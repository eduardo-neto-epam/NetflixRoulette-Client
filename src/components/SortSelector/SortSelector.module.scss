.container {
    color: var(--color-contrast);
    font-family: var(--font-main);
    font-size: 24px;
    font-weight: 400;
    position: absolute;
    top: 0;
    right: 0;
    margin-top: 10px;
    margin-right: 40px;

    ._inner {
        display: flex;
        width: 320px;
        justify-content: space-between;

        .label {
            color: var(--color-medium)
        }

        .selected-method {
            text-transform: uppercase;
            position: relative;
            cursor: pointer;

            &:after {
                content: '';
                width: 0; 
                height: 0;
                border-radius: 4px; 
                border-left: 12px solid transparent;
                border-right: 12px solid transparent;
                border-top: 12px solid var(--color-primary);
                position: absolute;
                top:50%;
                right: -40px;
                transform: translateY(-50%);
                pointer-events: all;
                cursor: pointer;
            }
        }
    }

    .options-container {
        transform: scaleY(0);
        transform-origin: top;
        background: var(--color-dark);
        position: absolute;
        top: 20px;
        right: -40px;
        text-align: right;
        animation: close .3s ease-in-out;
             
       &--show {
            animation: open .3s ease-in-out;
            transform: scaleY(1);

            @keyframes open {
                from {
                    transform: scaleY(0);
                    opacity: 0;
                }
                to {
                    transform: scaleY(1);
                    opacity: 1;
                }
            }
        }

        @keyframes close {
            from {
                transform: scaleY(1);
                opacity: 1;
            }
            to {
                transform: scaleY(0);
                opacity: 0;
            }
        }

        .option {
            list-style: none;
            padding: 10px;
            width: 200px;
            text-transform: uppercase;
            font-size: 18px;
            cursor: pointer;
        }

    }
}