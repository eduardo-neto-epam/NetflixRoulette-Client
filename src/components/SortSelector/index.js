import React, { useState } from 'react'
import styles from './SortSelector.module.scss'
import classnames from 'classnames/bind'

let cx = classnames.bind(styles)

const SortSelector = () => {

    const [ selectedMethod, setSelectedMethod ] = useState('release date')
    const [ viewOptions, setViewOptions ] = useState(false)

    const options = [
        'release date',
        'movie title',
        'movie genre',
    ]

    function handleSelection(e) {
        e.persist();
        setSelectedMethod(e.target.textContent)
        setViewOptions(prev => !prev)
    }

    function handleViewOptions() {
        setViewOptions(prev => !prev);
    }

    return (
        <div className={cx('container')}>
            <div className={cx('_inner')}>
                <div className={cx('label')}>SORT BY</div>
                <div
                 data-testid={`selected-method`}
                 className={cx('selected-method')}
                 onClick={handleViewOptions}
                 >
                     {selectedMethod}
                </div>
            </div>
           <ul className={cx(
               'options-container',
               {'options-container--show': viewOptions}
               )}>
                {options.map((option, index) => (
                    <li 
                        className={cx('option')}
                        key={index}
                        onClick={handleSelection}
                        >                
                        {option}
                    </li>
                ))}
            </ul>
        </div>
    )
}

export default SortSelector
