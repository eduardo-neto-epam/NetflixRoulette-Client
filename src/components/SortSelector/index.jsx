import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';
import { v4 as uuidv4 } from 'uuid';
import styles from './SortSelector.module.scss';

const cx = classnames.bind(styles);

const SortSelector = ({ selectedMethod, setSelectedMethod, options }) => {
  const [viewOptions, setViewOptions] = useState(false);

  function handleSelection(e) {
    setSelectedMethod(e.target.textContent);
    setViewOptions((prev) => !prev);
  }

  function handleViewOptions() {
    setViewOptions((prev) => !prev);
  }

  return (
    <div className={cx('container')}>
      <div className={cx('_inner')}>
        <div className={cx('label')}>SORT BY</div>
        <div
          data-testid="selected-method"
          className={cx('selected-method')}
          onClick={handleViewOptions}
          tabIndex="0"
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              handleViewOptions();
            }
          }}
          role="button"
        >
          {selectedMethod}
        </div>
      </div>
      <ul className={cx(
        'options-container',
        { 'options-container--show': viewOptions },
      )}
      >
        {options.map((option) => (
          <li
            className={cx('option')}
            key={`${option}${uuidv4()}`}
          >
            <button
              onClick={handleSelection}
              onKeyPress={(e) => {
                if (e.key === 'Enter') {
                  handleSelection(e);
                }
              }}
              type="button"
              tabIndex="0"
            >
              {option}
            </button>
          </li>
        ))}
      </ul>
    </div>
  );
};

SortSelector.propTypes = {
  selectedMethod: PropTypes.string.isRequired,
  setSelectedMethod: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default SortSelector;
