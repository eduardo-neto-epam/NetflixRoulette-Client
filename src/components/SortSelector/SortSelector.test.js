import React from 'react'
import ReactDOM  from 'react-dom'
import '@testing-library/react'
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { fireEvent, render, cleanup } from '@testing-library/react';
import SortSelector from './index'

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

afterEach(cleanup);

test('SortSelector renders correctly', () => {
    const tree = renderer
        .create(<SortSelector />)
        .toJSON();
    expect(tree).toMatchSnapshot();
  });


test('toggles options menu on click', () => {

function testClick() {
    act(() => { ReactDOM.render(<SortSelector />, container) });

    let clickableElement = container.querySelector('.selected-method');
    let displayedElement = container.querySelector('.options-container')

    fireEvent.click(clickableElement)
    expect(displayedElement.getAttribute('class')).toEqual('options-container options-container--show');

    fireEvent.click(clickableElement)
    expect(displayedElement.getAttribute('class')).toEqual('options-container');
    }
    
    testClick();
})

test('Option is selected when clicked', () => {

    function testClick() {
        const { getByTestId, getByText } = render(<SortSelector />);
    
        let clickableElement = getByText('movie title')
        let displayedElement = getByTestId('selected-method')
    
        fireEvent.click(clickableElement)
        expect(displayedElement.textContent).toEqual(clickableElement.textContent);
        }

    testClick();
})