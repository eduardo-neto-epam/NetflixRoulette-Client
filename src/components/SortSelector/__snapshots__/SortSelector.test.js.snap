// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortSelector renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="_inner"
  >
    <div
      className="label"
    >
      SORT BY
    </div>
    <div
      className="selected-method"
      data-testid="selected-method"
      onClick={[Function]}
    >
      release date
    </div>
  </div>
  <ul
    className="options-container"
  >
    <li
      className="option"
      onClick={[Function]}
    >
      release date
    </li>
    <li
      className="option"
      onClick={[Function]}
    >
      movie title
    </li>
    <li
      className="option"
      onClick={[Function]}
    >
      movie genre
    </li>
  </ul>
</div>
`;
